source/dem/dem.cc

```diff
@@ -1065,10 +1066,11 @@ DEMSolver<dim>::solve()
 
       // Write checkpoint if needed
       if (parameters.restart.checkpoint &&
-          simulation_control->get_step_number() %
-              parameters.restart.frequency ==
-            0)
+          checkpoint_controller.is_checkpoint_time_step(
+            simulation_control->get_step_number()))
         {
+          checkpoint_controller.increment_checkpoint_id();
+
```

Comment:
I don't this this is a good design.
Why not increment the checkpoint id from within the is_checkpoint_time_step?
In the end you should only check ONCE per time-step taht this is a checkpoint time step and so you could do the incrementation from within the check.

I don't like the idea of controlling the CheckpointController from the outside. The checkpoint controller should be independent in terms of logic from teh simulation.